modules/currencies/src/lib.rs

```diff
@@ -637,12 +657,274 @@ impl<T: Config> MultiReservableCurrency<T::AccountId> for Pallet<T> {
+			}
+			_ => <T::MultiCurrency as fungibles::Inspect<_>>::reducible_balance(asset_id, who, keep_alive),
+		}
+	}
+
+	fn can_deposit(asset_id: Self::AssetId, who: &T::AccountId, amount: Self::Balance) -> DepositConsequence {
+		match asset_id {
+			CurrencyId::Erc20(_) => {
+				if amount.is_zero() {
+					return DepositConsequence::Success;
+				}
+
+				if <Self as fungibles::Inspect<_>>::total_issuance(asset_id)
+					.checked_add(&amount)
+					.is_none()
+				{
+					return DepositConsequence::Overflow;
+				}
+
+				if <Self as fungibles::Inspect<_>>::balance(asset_id, who).saturating_add(amount)
+					< <Self as fungibles::Inspect<_>>::minimum_balance(asset_id)
+				{
+					return DepositConsequence::BelowMinimum;
+				}
+
+				DepositConsequence::Success
+			}
+			id if id == T::GetNativeCurrencyId::get() => {
+				<T::NativeCurrency as fungible::Inspect<_>>::can_deposit(who, amount)
+			}
+			_ => <T::MultiCurrency as fungibles::Inspect<_>>::can_deposit(asset_id, who, amount),
+		}
+	}
+
+	fn can_withdraw(
+		asset_id: Self::AssetId,
+		who: &T::AccountId,
+		amount: Self::Balance,
+	) -> WithdrawConsequence<Self::Balance> {
+		match asset_id {
+			CurrencyId::Erc20(_) => match <Self as MultiCurrency<_>>::ensure_can_withdraw(asset_id, who, amount) {
+				Ok(()) => WithdrawConsequence::Success,
+				_ => WithdrawConsequence::NoFunds,
+			},
+			id if id == T::GetNativeCurrencyId::get() => {
+				<T::NativeCurrency as fungible::Inspect<_>>::can_withdraw(who, amount)
+			}
+			_ => <T::MultiCurrency as fungibles::Inspect<_>>::can_withdraw(asset_id, who, amount),
+		}
+	}
+}
+
+impl<T: Config> fungibles::Mutate<T::AccountId> for Pallet<T> {
+	fn mint_into(asset_id: Self::AssetId, who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
+		<Self as MultiCurrency<_>>::deposit(asset_id, who, amount)
+	}
+
+	fn burn_from(
+		asset_id: Self::AssetId,
+		who: &T::AccountId,
+		amount: Self::Balance,
+	) -> Result<Self::Balance, DispatchError> {
+		if amount.is_zero() {
+			return Ok(amount);
+		}
+
+		let actual = match asset_id {
+			CurrencyId::Erc20(_) => return Err(Error::<T>::Erc20InvalidOperation.into()),
+			id if id == T::GetNativeCurrencyId::get() => {
+				<T::NativeCurrency as fungible::Mutate<_>>::burn_from(who, amount)
+			}
+			_ => <T::MultiCurrency as fungibles::Mutate<_>>::burn_from(asset_id, who, amount),
+		}?;
+		Self::deposit_event(Event::Withdrawn {
+			currency_id: asset_id,
+			who: who.clone(),
+			amount: actual,
+		});
+		Ok(actual)
+	}
+}
+
+impl<T: Config> fungibles::Transfer<T::AccountId> for Pallet<T> {
+	fn transfer(
+		asset_id: Self::AssetId,
+		source: &T::AccountId,
+		dest: &T::AccountId,
+		amount: Self::Balance,
+		keep_alive: bool,
+	) -> Result<Self::Balance, DispatchError> {
+		if amount.is_zero() || source == dest {
+			return Ok(amount);
+		}
+
+		let transferred_amount = match asset_id {
+			CurrencyId::Erc20(_) => {
+				<Self as MultiCurrency<_>>::transfer(asset_id, source, dest, amount)?;
```

Comment:
This will end up with depositing `Transferred` event twice.

Reply:
I think the `transfer` impl can be delegated to `<Self as MultiCurrency<_>>::transfer`. They are the same except return types.

Reply:
Yeah i had a look - it's the keepalive parameter that's making it difficult. I wanted to keep this so the keepalive can be passed into the trait bounded implementation.

Reply:
I'll move the deposit event to the inside of the Erc20 match branch so we don't double deposit as you said

Reply:
Then you will need to deposit events in the last match branch, `orml-tokens`'s `MultiCurrency` impl has no events deposit.

Reply:
Addressed here.
Also tidied up events in moduke-currencies and unit tests
https://github.com/open-web3-stack/open-runtime-module-library/pull/729